### ISSUE TITLE
Proxy http2 client

### ISSUE THREAD

**Issue by gofrailo** [h2o/h2o#170]
Hi,

Correct me if I am wrong but there could be major speed gain if the proxy could also support http2 to backends. Do you have any plans to have proxy support for this?

Thanks a lot,

**Comment by kazuho** [comment]
@gofrailo Thank you for starting the discussion on the topic.

Whether we would see speed gain by supporting HTTP/2 from the proxy module to upstream depends on the network latency between H2O and the upstream.

At the moment, the primary focus of the proxy module is to provide a way to connect to application servers running in the same data center as H2O is running.  For such use case there would be only marginal merit in supporting HTTP/2 for backend connections, esp. considering the fact that application servers tend to be implemented as blocking servers (e.g. non-event driven) which means that they are likely to stay in using HTTP/1 for the time being (relates to #64).

Therefore I consider the topic as a low priority issue.

**Comment by HoneyryderChuck** [comment]
I think your reasoning has merit for the legacy application servers around, but not 2 years after wide HTTP2 deployment. New implementations pop up, and it still makes sense to put them behind a proxy. A non- HTTP module that passes a marshalled request ( think thrift or zero mq?) Would also make sense. 

I'd very much appreciate HTTP2 proxying :) or are other servers around that already do it?

**Comment by yannick** [comment]
@HoneyryderChuck  for a prototype you could hack it together by compiling in mruby http2 clients, but that will give you way worse performance. i once tried that but its useless, so not recommended!

also as kazuho mentioned, as long as your backend does not send useless headers you will see no speed benefits over http/1.1. or why would you, as long as the tcp connection stays open?

RPC backends could be interesting but should be implemented via mruby first (look at the redis PR, you could do something similar for thrift/protobuf etc). then again also there, you would not suddenly see any magic performance gains, it will actually be slower for 1:1 proxying cases!

**Comment by HoneyryderChuck** [comment]
@yannick I agree with all of your points. The main goal is interoperability, not performance. If you're hacking an application server which also supports HTTP/1.1 and HTTP/2, you'd want to serve as many HTTP/2 traffic as possible and get rid of the legacy. @kazuho's arguments made total sense in 2015 when adoption was lower and you needed to downgrade versions on flight for legacy applications, but same as IPv4-to-6-and-vice-versa proxies, you'd like to see the legacy version vanish. If I deploy a go server behind a proxy today (not just with h2o), will it only be served HTTP/1.1 (and go doesn't support unencrypted HTTP/2)?

If we're talking same-datacenter upstream, one might not see a big benefit in TCP connection usage (there's more tolerance for HTTP head-of-line blocking, but doesn't erase it), but what about CDN-to-upstream-outside? How about the parse-and-reparse of HTTP requests/responses? (although, RPC backend would be better suited). One can say there's no apparent speed benefit, but are there reasons to anticipate a performance downgrade?

Thx for support and the tip on the redis PR, I'll look into it and check the lib internals (I've only interacted with the server). Maybe my arguments can be better served by a prototype.
